### ISSUE TITLE
Animationtree blending issue 

### ISSUE THREAD

**Issue by Kliefer** [godotengine/godot#81021]
### Godot version

v4.1.1.stable.official [bd6af8e0e]

### System information

Godot v4.1.1.stable - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 4070 Ti (NVIDIA; 31.0.15.3640) - Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz (8 Threads)

### Issue description

In our 3D project we have a certain problem with the animation tree blending. We saw some issues similar to ours but not that it is the same in our opinion.
We are using .dae for the animation and .fbx for the skinned mesh because the import just deletes unused tracks in the animation which we will also write a more detailed version for this issue.
While blending animation via the animation tree we found a problem that if you have an aniomation that is 178° and you want to blend it to an animation that ist for example -178° the entire bone is flipping in the other direction all the way around. So instead of rotating 4° it rotates instead 360°
Other thing to note is that it seems that the animationplayer blends the animations diffrently than the animationtree because we didnt encounter any of the listed problems there atleast for smaller angle flips.

We also tested this in our animation software namly Autodesk Maya (2019 -2024) and other Game engines which are Unity and Unreal see the video attachments below

https://github.com/godotengine/godot/assets/26251802/a71fcc07-dc15-4cd6-bc23-c07d81708ddf

The Bug in Godot the flickering of the weapon is the issue 

https://github.com/godotengine/godot/assets/26251802/cd4d2060-3ff7-4cc9-aeeb-ff47437e2831

https://github.com/godotengine/godot/assets/26251802/8f178071-8879-43b6-9c83-ce31f645e9b5

https://github.com/godotengine/godot/assets/26251802/ed3104d5-a703-4b52-92d3-90c81cbcd843

a more extrem example with a 360° animation which unity  can handle it without a problem

https://github.com/godotengine/godot/assets/26251802/b91877d4-2ea2-43b5-af8b-be26a12ea9c7

The same animation in Unreal

### Steps to reproduce

- Create two animations and both are close 180 in our one animation is ca. 179° and the one to blend to is  ca.182°.
- Create an Animationtree and load both animation in inside the animationtree statemachine
- Connect animation A and B, then set a blend value of 0,1 (that a value is present)
- Play the animation with the blend and see how the angle and the animation is flipping

### Minimal reproduction project

[Testproject.zip](https://github.com/godotengine/godot/files/12446528/Testproject.zip)
To replicate the issue go to the animationtree and blend between the animations

**Comment by TokageItLab** [comment]
Probably the BoneRest of the weapon is incorrect. Bone select a path that is closer from the rest in AnimationTree's brending.

For example, when blending a high kick with a back kick, the leg must avoid to a fracture where the leg blends through the body, so the leg BoneRest should be on the ground.

**Comment by Kliefer** [comment]
The BoneRest of the Weapon Bone is on 0,0,0 under the Rootbone and other engines had no problems with the animationblend. 
We had a problem with script with a similar behaviour and we fixed it with this function :

https://ask.godotengine.org/27625/equivalent-of-gamemakerss-function-angle_difference?show=27631#a27631

Maybe that could help you with quaternion interpolation so that there is no longer a long quaternion artifact

**Comment by TokageItLab** [comment]
Other engines are supposed to treat Humanoid as a special bone, but Godot treats them no differently, so BoneRest is important. Or it is possible that other engines do not interpolate high kicks and back kicks well.

**Comment by Kliefer** [comment]
![image](https://github.com/godotengine/godot/assets/26251802/d78c50ed-4f25-43be-aacc-e7a59db347db)

We already usedBoneAttachments and parenting the weapon bone to the hand yielded the same results sadly.

But thank you for looking at this problem and taking the time with it.

**Comment by TokageItLab** [comment]
Your R_Weapon_Hand Rest:
![image](https://github.com/godotengine/godot/assets/61938263/712e6abf-7267-43b0-beeb-1cf271e9d1bb)

But in walk:
![image](https://github.com/godotengine/godot/assets/61938263/8f2a55b6-df40-4649-bcb8-953066df0f8e)

So your R_Weapon_Hand Rest needs to face back like:
![image](https://github.com/godotengine/godot/assets/61938263/0ab5e3f3-9b0a-494a-b58b-eaf9a223dded)

Fixed project:
[Testproject.zip](https://github.com/godotengine/godot/files/12490945/Testproject.zip)

**Comment by jcarlosrc** [comment]
It seems devs do not want to address this issue. It's gotten unnecessarily complex to make basic animations to deal with the way Godot blends them. They cannot be just imported. The animators need to design them with the way Godot manages them in mind. Not good.

**Comment by TokageItLab** [comment]
As previously mentioned in #81755, the shortest path interpolation based on the current animation has the following serious unusable problems

> Unpredictable rotations: It does not make sense to get the shortest path from the prev frame. Even if it is the shortest path when comparing blended animations' current frame, it may will be the not shortest path by progression of the animation during blending. At the moment it is no longer the shortest path, the blending will cause an unintended flip.

Probably any DCC software, not just Godot, will have to choose between that inconvenient problem or a solution based on rest.

**Comment by MatthewBlanchard** [comment]
> As previously mentioned in [#81755 (comment)](https://github.com/godotengine/godot/issues/81755#issuecomment-1722317059), the shortest path interpolation based on the current animation has the following serious unusable problems:
> 
> > Unpredictable rotations: It does not make sense to get the shortest path from the prev frame. Even if it is the shortest path when comparing blended animations' current frame, it may will be the not shortest path by progression of the animation during blending. At the moment it is no longer the shortest path, the blending will cause an unintended flip.
> 
> Moreover, if we want to do it with Godot, the blending result depends on the iterating order, meaning that the same amount of blending will change depending on the order in which the nodes are connected.
> 
> Probably any DCC software, not just Godot, will have to choose between that using current frame interpolation with unusable problem or a deterministic solution based on rest.
> 
> Also, if using solution with rest basis, there is a workaround to rewrite rest depending on state as mentioned in [#88805 (comment)](https://github.com/godotengine/godot/issues/88805#issuecomment-1968831771). It is the programmer's role, so there is no need to mind about this on the animator's side.

Man I don't know how you can say this with a straight face. Every other engine in this example just worked when they imported the animation. Your continued closing of these issues and recommendations of jank fixes for BASIC animations is absurd. Can you please just mark ONE of these issues as an actual bug and leave it open?

**Comment by TokageItLab** [comment]
As I mentioned above, the only thing we can deal with here is how to set up REST. Changing the underlying blending algorithm should never happen.

Most of these problems have been reported when **users import models from outside without being aware of rest and without using retarget**.

Not a single problem has yet been submitted when compared to **Unity or Unreal**. Also, since those game engines implicitly do retargeting, comparisons with godot should be made with cases that use Godot's retargeting feature.

As I said above, simple on-the-fly slerp may be adopted in DCC software where the blend tree is not expected to be large, but the game engine may accumulate a number of rotations, which can easily break the user's intended blending.

**Comment by jcarlosrc** [comment]
It is really amazing how basic usability is being taken out of the engine to solve complex technical cases. Making a basic blending for an FPS weapon system where the hands need to be attached to the weapon in a consistent way is a complete mess. It seems the user needs to guess what is the suitable rest ppose. Finding such suitable rest pose is almost impossible in practice in some cases. Moreover, it is suggested that the rest poses needs to be changed in code dynamically depending on the animations and that is the developer responsibility. Also, more extreme rotations need to use root motion bone. This makes  blending basic animations a completely unnatural and unacceptable hard task. While comparisons are ugly, it is undeniable that other engines do not require such strange amount of work to show a proper blending. Maybe complex set ups require more attention but this is preferable. Simple blending cases should not require such amounts of manual guessing and work. In some cases it is just too hard to be achievable. This is not right. This is not an improvement or "for better blending".

**Comment by TokageItLab** [comment]
@jcarlosrc As I have said many times, the problem with this issue is that the user is not using Retarget, and the only reason other game engines do not have the problem is that retargeting is only halfway enforced. Your comparisons that do not assume retargeting is completely meaningless.

It will be interesting to see how other game engines solve the problem of rest-dependent slerp, but at least there should be a trade-off with the problem indicated in https://github.com/godotengine/godot-proposals/issues/9215#issuecomment-2081532539.

**Comment by RudyFisher7** [comment]
> [@jcarlosrc](https://github.com/jcarlosrc) As I have said many times, the cause of this issue is that the user is not using retarget, and the only reason other game engines do not have the problem is that retargeting is almost enforced. Your comparisons that do not assume retargeting is completely meaningless.

> Also, no one has been able to provide an actual other game engine's example that prioritizes the shortest rotation without breaking the blend of three or more animations.

To translate what this dev says here in layman's terms: "Unreal and Unity don't break because they work differently than Godot. Your comparison of how much easier those engines are to use in this case is invalid because the engines work differently." But the whole point of this issue is the OP wants it to WORK DIFFERENTLY SO THAT IT IS EASIER TO USE. 

Ay ay ay... smh. Do better.
